Fixes to make bootstrapping from Linux/musl work

valpackett opened the pull request:
Here's a bunch of fixes that allowed me to build FreeBSD under @chimera-linux which uses musl libc and a port of our own core programs. The more complex changes are definitely subject to improvements, suggestions/bikeshedding welcome!

+also required: https://github.com/openzfs/zfs/pull/15780

bsdimp commented:
This looks quite interesting... i like parts and have concerns about parts. Ill go through it with a fine toothed comb.

emaste reviewed (commented):
style(9) disagrees with this change, `(<sys/param.h> includes <sys/types.h>; do not include both.)` so we'll need to address that somehow -- possibly by amending style to include something like `do not include both, unless required for portability`.

(gh code review does not make it clear the code review comment is going to be left on the general conversation, not the specific file or commit -- this is with respect to 032c239e83ef2917921153ecb5daa6ab1abd4056 sys/capsicum.h)

emaste commented on kerberos5/include/config.h:
```diff
@@ -694,7 +696,9 @@
 #define HAVE_RES_NDESTROY 1
 
 /* Define to 1 if you have the `res_nsearch' function. */
+#if __RES >= 19991006
```
cc @cschuber - is there any automation involved in the generation of config.h on krb5 updates?

bsdimp replied:
This likely needs a comment, but this doesn't feel like the right place to solve this...

emaste commented:
jevents.c should coordinate with upstream

jrtc27 commented on tools/build/cross-build/include/common/langinfo.h:
```diff
@@ -0,0 +1,46 @@
+
+#include_next <langinfo.h>
+
+#ifndef ALTMON_1
```
https://github.com/freebsd/freebsd-src/commit/bd234c0d4c8256db7e5a1fdda9ef311c9e0080e4 should have made these no longer required (they aren't present on macOS yet things build just fine today). Are you sure you need them, and if so, why?

valpackett replied:
I've definitely started working on this before pulling that commit in. Works fine without the file now, I'll remove it.

bsdimp commented:
> style(9) disagrees with this change, `(<sys/param.h> includes <sys/types.h>; do not include both.)` so we'll need to address that somehow -- possibly by amending style to include somethin


We need to document nonstandard usage.

I have a lot of other comments as well. Will do them after bowling.

bsdimp reviewed (changes requested):
Please don't land anything else from this PR. I have several comments pending i wanted to think through.

bsdimp commented:
```
diff --git a/tools/build/cross-build/include/linux/sys/types.h b/tools/build/cross-build/include/linux/sys/types.h
index 770b3057f8a8..e72e1419bccd 100644
--- a/tools/build/cross-build/include/linux/sys/types.h
+++ b/tools/build/cross-build/include/linux/sys/types.h
@@ -35,6 +35,11 @@
  */
 #pragma once
 
+/*
+ * FreeBSD's types.h always includes sys/cdefs.h, but Linux headers often don't.
+ */
+#include <sys/cdefs.h>
+
 #include_next <sys/types.h>
 
 #ifndef __size_t
```
I think will fix the issue...

bsdimp commented on contrib/mandoc/config.h:
```diff
@@ -13,7 +13,7 @@
 #define HAVE_ERR 1
 #define HAVE_FTS 1
-#ifdef __GLIBC__
+#ifdef __linux__
```
This likely should be upstraemed, but it changes a local change, so it's OK.Both the changes need to go up.... :)

bsdimp commented on include/nl_types.h:
```diff
@@ -34,6 +34,7 @@
 #ifndef _NL_TYPES_H_
 #define _NL_TYPES_H_
 
+#include <sys/cdefs.h>
```
I commented on this.

bsdimp commented on lib/libpmc/pmu-events/jevents.c:
```diff
@@ -1355,7 +1355,7 @@ int main(int argc, char *argv[])
 
 static int
-#ifdef __GLIBC__
+#ifdef __linux__
```
this looks good. __GLIBC__ vs __linux__ seems sane.

bsdimp commented on sys/sys/capsicum.h:
```diff
@@ -40,6 +40,7 @@
 #define	_SYS_CAPSICUM_H_
 
 #include <sys/param.h>
+#include <sys/types.h>
```
This should likely just be added to `./tools/build/cross-build/include/linux/sys/param.h`

bsdimp commented on usr.bin/grep/grep.c:
```diff
@@ -62,9 +62,8 @@ const char	*errstr[] = {
-/* Flags passed to regcomp() and regexec() */
+/* Flags passed to regcomp() */
 int		 cflags = REG_NOSUB | REG_NEWLINE;
-int		 eflags = REG_STARTEND;
```
This is a very odd change.... It should be #ifdef linux or something, no?

valpackett replied:
There's `#ifdef REG_STARTEND` but inside of `util.c`. There it's choosing between using the flag if it exists, and a fallback manual implementation in the `#else`. With that, the global default for the flags doesn't quite make sense.

bsdimp reviewed (commented):
Mostly I'm wondering why the sed/grep changes aren't conditional.

bsdimp commented:
Also, if the changes to use cross-tools includes work... then maybe we can eliminate some of the added ctypes.h includes.

arichardson commented on tools/build/cross-build/include/common/pwd.h:
```diff
@@ -51,6 +51,7 @@ int uid_from_user(const char *name, uid_t *uid);
 const char *user_from_uid(uid_t uid, int noname);
 
 #ifdef __linux__
+#include <sys/cdefs.h>
```
I guess this is just for the __unused? Could add a cast to void inside the function body instead?

bsdimp commented:
Ping? What's up with this? Sounded like you were preparing an updated, but it doesn't look like one's been pushed.
I don't have this environment, so I can't easily mine this for changes and commit piecemeal.
Can you rebase and update the PR please? Despite all the fiddly comments, I think this would be good to have.
If the upstream bits are getting in the way, maybe just resubmit those w/o them, we'll get that in and then see what
to do about upstream?
Thanks!

valpackett commented:
Oop… yeah, sorry, was so busy with all the everything again. Thanks for the ping. Rebased, removed langinfo, did the cdefs-via-types thing instead of touching capsicum.h.

BTW, feel free to do more cherry-picking-and-pushing of commits you'd consider uncontroversial, the PR getting smaller always makes it better to work with (:

bsdimp commented on tools/build/cross-build/include/common/sys/_types.h:
```diff
@@ -45,3 +45,6 @@
 typedef __builtin_va_list __va_list;
+
+/* Needed for opensolaris compat. */
+typedef	__int64_t	off64_t;
```
this is fine now, but may cause some grief in the future when I land our better compat with open solaris off64_t stuff

bsdimp commented on tools/build/cross-build/include/mac/sys/_types.h:
```diff
@@ -41,5 +41,3 @@
  */
 typedef __darwin_ct_rune_t __ct_rune_t;
-/* Needed for opensolaris compat. */
-typedef	__int64_t	off64_t;
```
I don't understand this one...

jrtc27 replied:
Moved to common given it’s needed for musl, it seems

valpackett replied:
Exactly

bsdimp commented on tools/build/cross-build/progname.c:
```diff
@@ -36,7 +36,6 @@
 #include <sys/cdefs.h>
 #include <stdlib.h>
 
-#ifdef __GLIBC__
```
or this

jrtc27 replied:
I guess because tools/build/Makefile only builds it on Linux, so it was previously a stupid ifdef, and now is unhelpful for musl.

arichardson replied:
I feel like this file might have been compiled for macos in the past so the ifdef was there to only use it for Linux. Dropping it looks good.

bsdimp commented on usr.bin/localedef/bootstrap/ctype.h:
```diff
@@ -44,13 +44,23 @@
-#ifndef isxdigit
 __BEGIN_DECLS
+#ifndef isalpha
 int	isalpha(int);
```
quick question here: there's two ways to fix this: one is how you've done (and is likely fine for bootstrap), the other is to make these int (isalpha)(int); to prevent the macro expansion. Is there a reason to prefer one over the other?

jrtc27 replied:
I guess you shouldn't have a prototype for something that's implemented as a macro, since it doesn't really exist

bsdimp reviewed (commented):
just a few notes... I'll see what I can land from this and we'll iterate.

arichardson commented on usr.bin/sed/process.c:
```diff
@@ -672,11 +671,23 @@ regexec_e(regex_t *preg, const char *string, int eflags, int nomatch,
 		defpreg = preg;
 
 	/* Set anchors */
+#ifndef REG_STARTEND /* in cross-build tools */
```
This looks different from the wrapper in grep.c (the former does not have a loop). Would it make sense to pull this out into a shared helper?

jrtc27 commented on tools/build/mk/Makefile.boot:
```diff
@@ -41,15 +41,15 @@ CWARNFLAGS.clang+=-Wno-typedef-redefinition
 CFLAGS+=	-D_GNU_SOURCE=1
 # Needed for sem_init, etc. on Linux (used by usr.bin/sort)
 LDADD+=	-pthread
+.if exists(/usr/lib/libfts.so) || exists(/usr/lib/libfts.a) || exists(/lib/libfts.so) || exists(/lib/libfts.a)
```
This feels dodgy

jrtc27 reviewed (commented):
My general comment is that there's very little by way of explanation and justification, with a lot of unrelated changes being lumped into a single commit that has no body in the message. Each change should document *why* it's needed, and be limited to one thing. As it stands for a lot of these I look at the diff and can only really speculate on why they're needed.

bsdimp commented:
> @bsdimp Please liaise with me before landing any of this, I would like to ensure I review it first

Will do. The changes themselves likely are OK, but I share your concerns about justification and regressions, though the cross build github jobs work. The changes also aren't chunked up quite right, and need to be regroups a little. I'll post phab reviews and CC you.  I need to find a musl build env...

bsdimp commented on contrib/mandoc/config.h:
```diff
@@ -13,7 +13,7 @@
 #define HAVE_ERR 1
 #define HAVE_FTS 1
-#if defined(__GLIBC__) || defined(__APPLE__)
+#if defined(__linux__) || defined(__APPLE__)
```
Is this really #ifndef __FreeBSD__? Or does NetBSD/OpenBSD also have this new const interface?

bsdimp commented:
OK. To make progress on this, I've setup a Alpine Linux VM. I've started mining this PR for bits of code that can be committed. I've also added a few comments for some weird things that I've noticed. I'm testing this now. https://reviews.freebsd.org/D45349 is the first of 7 reviews that is the result. These are all uninteresting bits. I added Val as a co-author on almost all of them. I could swap author and co-author on them though. I wrote either extra comments, or a commit message for these, so I think it's only fair to share the credit and I'm agnostic on how. The co-authored-by stuff was the path of least resistance for me, but let's chat if that bothers anybody. It also slices up what's here a bit differently, but not that differently. I hope Co-authored-by: is the right git trailer to use for github to give co-credit. If not, I'll change them to whatever does.

I added @jrtc27 to the reviews as well.  My hope is that we can commit those reviews, close this review out and then do the rest of this pull request with new, smaller pull requests. The size of this request made it linger far too long and I'd like to try to get things in faster, which means I think that future pull requests should be smaller and more easy to manage. There's just too many moving parts otherwise.

bsdimp commented:
OK. All three builds succeeded in the musl branch I pushed. Though on alpine Linux 'success' was defined to be 'made it as far as the hack and slash version I did to validate what I cleaned up and put into the musl branch'. The stat stuff looks ready too, I overlooked it earlier and I'm out of time today. The base64 stuff, the resolver stuff, the pthreads stuff and the regexp stuff I don't pretend to grok all remain, each of which may need some tweaking and I'd suggest should be a separate future pull request in the future to keep things from getting bogged down.

bsdimp commented:
so, it's time to close this one. I've mined all the best bits out. Maybe there's more to get (like the stat.h patch), but I'm going to close this and we'll start on the other stuff from that base.
